pkg/services/live/push/push.go

```diff
@@ -58,25 +80,97 @@ func (g *Gateway) IsEnabled() bool {
 	return g.Cfg.IsLiveEnabled() // turn on when Live on for now.
 }
 
-func (g *Gateway) Handle(ctx *models.ReqContext) {
-	streamID := ctx.Params(":streamId")
+func extractKey(req *http.Request) string {
+	header := req.Header.Get("Authorization")
+	parts := strings.SplitN(header, " ", 2)
+	var keyString string
+	if len(parts) == 2 && parts[0] == "Bearer" {
+		keyString = parts[1]
+	}
+	return keyString
+}
+
+func (g *Gateway) isAuthenticated(req *http.Request) (bool, error) {
+	keyString := extractKey(req)
+	if keyString == "" {
+		return false, nil
+	}
+	g.keyMu.RLock()
+	existingKey, ok := g.keyCache[keyString]
+	g.keyMu.RUnlock()
+	if !ok {
+		// TODO: share the code below with middleware package.
+		decoded, err := apikeygen.Decode(keyString)
+		if err != nil {
+			return false, nil
+		}
+		keyQuery := models.GetApiKeyByNameQuery{KeyName: decoded.Name, OrgId: decoded.OrgId}
+		if err := bus.Dispatch(&keyQuery); err != nil {
+			if errors.Is(err, models.ErrInvalidApiKey) {
+				return false, nil
+			}
+			return false, err
+		}
+		apikey := keyQuery.Result
+		// validate api key.
+		isValid, err := apikeygen.IsValid(decoded, apikey.Key)
+		if err != nil {
+			return false, nil
+		}
+		if !isValid {
+			return false, nil
+		}
 
+		now := time.Now()
+
+		if apikey.Expires != nil && *apikey.Expires <= now.Unix() {
+			return false, nil
+		}
+		g.keyMu.Lock()
+		g.keyCache[keyString] = CachedKey{CheckedAt: now, IsValid: true}
+		g.keyMu.Unlock()
+	} else {
+		if !existingKey.IsValid {
```

Comment:
Feels like this case will never happen or I might missing something. You probably need to store when api key expires in the cache to be able to invalidate it - otherwise Grafana would allow requests for a cached api key that in theory has expired.

Reply:
Yep - there is a comment about this, I have not implemented cache invalidation routine yet since wanted to get a green light with general idea about bypassing Macaron and caching keys.